docs/tutorials_and_cookbooks/agents_with_tools.md

```diff
@@ -0,0 +1,304 @@
+assistant_model_config = ChatGPTConfig(
+    tools=tool_list,
+    temperature=0.0,
+)
+
+# Set the agent
+agent = ChatAgent(
+    assistant_sys_msg,
+    model=model,
+    tools=tool_list
+)
+```
+
+Let's define two test prompt for the agent to run
+```python
+# Set prompt for the search task
+prompt_search = ("""When was University of Oxford set up""")
+# Set prompt for the calculation task
+prompt_calculate = ("""Assume now is 2024 in the Gregorian calendar, University of Oxford was set up in 1096, estimate the current age of University of Oxford""")
+
+# Convert the two prompt as message that can be accepted by the Agent
+user_msg_search = BaseMessage.make_user_message(role_name="User", content=prompt_search)
+user_msg_calculate = BaseMessage.make_user_message(role_name="User", content=prompt_calculate)
+
+# Get response
+assistant_response_search = agent.step(user_msg_search)
+assistant_response_calculate = agent.step(user_msg_calculate)
+```
+
+Let's see how did the agent use the tool for the 2 tasks
+```python
+print(assistant_response_search.info['tool_calls'])
+```
+
+```markdown
+>>> 
+[FunctionCallingRecord(func_name='search_wiki', args={'entity': 'University of Oxford'}, result="The University of Oxford is a collegiate research university in Oxford, United Kingdom. There is evidence of teaching as early as 1096, making it the oldest university in the English-speaking world and the world's second-oldest university in continuous operation. It grew rapidly from 1167, when Henry II banned English students from attending the University of Paris. After disputes between students and Oxford townsfolk in 1209, some academics fled north-east to Cambridge where they established what became the University of Cambridge. The two English ancient universities share many common features and are jointly referred to as Oxbridge.")]
+```
+
+```python
+print(assistant_response_calculate.info['tool_calls'])
+```
+
+```markdown
+>>> 
+[FunctionCallingRecord(func_name='sub', args={'a': 2024, 'b': 1096}, result=928)]
+```
+
+Now the single agent has the ability to use tool, let set one small AI society, the society would include one user agent and one assistant agent, the assistant agent is the one we just set can use tool.
+```python
+from camel.societies import RolePlaying
+from camel.agents.chat_agent import FunctionCallingRecord
+from camel.utils import print_text_animated
+from colorama import Fore
+
+# Set a task
+task_prompt=("Assume now is 2024 in the Gregorian calendar, "
+        "estimate the current age of University of Oxford "
+        "and then add 10 more years to this age.")
+
+# Set role playing
+role_play_session = RolePlaying(
+    assistant_role_name="Searcher",
+    user_role_name="Professor",
+    assistant_agent_kwargs=dict(
+        model=ModelFactory.create(
+            model_platform=ModelPlatformType.OPENAI,
+            model_type=ModelType.GPT_3_5_TURBO,
+            model_config_dict=assistant_model_config.__dict__,
```

Comment:
all of `model_config.__dict__` should be change to `model_config.as_dict()`,
as shown in the pydantic PR.

Reply:
since pydantic refactor hasn't been merged, how about do the update in the pydantic PR later